### DISCUSSION TITLE
Consistency model of JetStream

### DISCUSSION THREAD

**Discussion by tekumara** [nats-io/nats-server#4577] [Q&A] [answered]
Which of the [formal consistency models](https://jepsen.io/consistency) does JetStream offer for messages in a stream? 

[The docs](https://docs.nats.io/nats-concepts/jetstream#:~:text=JetStream%20uses%20a%20NATS%20optimized%20RAFT%20distributed%20quorum%20algorithm%20to%20distribute%20the%20persistence%20service%20between%20nats%20servers%20in%20a%20cluster%20while%20maintaining%20immediate%20consistency%20even%20in%20the%20face%20of%20Byzantine%20failures) mention immediate consistency:

> JetStream uses a NATS optimized RAFT distributed quorum algorithm to distribute the persistence service between nats servers in a cluster while maintaining immediate consistency even in the face of Byzantine failures.

But I'm not sure which of the formal consistency models the NATS optimized RAFT distributed quorum algorithm represents.

**Comment by jnmoyne** [comment, marked as answer]
When the doc mentions immediate consistency, it is in contrast to [eventual consistency](https://en.wikipedia.org/wiki/Eventual_consistency). It is about how 'writes' (i.e. publishing a message to a stream).

JetStream is an immediately consistent distributed storage system in that every new message stored in the stream is done so in a unique order (when those messages reach the stream leader) and that the acknowledgement that the storing of the message has been successful only happens as the result of a RAFT vote between the NATS JetStream servers (e.g. 3 of them if replicas=3)  handling the stream. 

This means that when a publishing application receives the positive acknowledgement to it's publication to the stream you are guaranteed that everyone will see that new message in their updates _in the same order_ (and with the same sequence number and time stamp).

This 'non-eventual' consistency is what enables 'compare and set' (i.e. compare and publish to a stream) operations on streams: because there can only be one new message added to a stream at a time.

To map back to those formal consistency models it means that for writes, NATS JetStream is [Linearizable](https://jepsen.io/consistency/models/linearizable).

On the read side the models don't really apply because JetStream doesn't support atomic batching of multiple operations together so the only kind of 'transaction' is the persisting, replicating and voting of a single operation on the stream, but in essence JetStream is [serializable](https://jepsen.io/consistency/models/serializable) because messages are added to a stream in one global order (which you can control using compare and publish).

**Comment by derekcollison** [comment]
Ack with sync option and AckAll ack policy does provide multiple messages to be all acked at once.
